File: lab3/python/darray.py
from enum import Enum
import config
import sys

class darray:
    def __init__(self):
        self.array = []
        self.sorted = False
        
    def insert(self, string):
        self.array.append(string)
        
        # changing the array means it may no longer be sorted
        self.sorted = False
        
    def find(self, value):
        if (config.mode == SearchModes.LINEAR_SEARCH.value):
            for string in self.array:
                if (value == string):
                    return True
            return False
        else:
            if (not self.sorted):
                if (config.verbose > 0):
                    print("Dynamic Array not sorted, sorting... \n")
                    
                self.sort(config.mode)
                if (config.verbose > 0):
                    print("Dynamic Array sorted\n")
                
                self.sorted = True
            first = 0
            last = len(self.array) - 1
            middle = int((first + last)/2)
            
            while (first <= last):
                if (self.array[middle] < value):
                    first = middle + 1
                elif (self.array[middle] == value):
                    return True
                else:
                    last = middle - 1
                middle = int((first + last)/2)
                
        return False
        
    def print_set(self):
        print("DArray:\n")
        for i in range(len(self.array)):
            print("\t%s\n" % self.array[i])
            
    def print_stats(self):
        print("Dynamic array contains %d elements\n" % len(self.array))
        
    def sort(self, select):
        if (select == SearchModes.BINARY_SEARCH_ONE.value):
            self.insertion_sort()
        elif (select == SearchModes.BINARY_SEARCH_TWO.value):
            self.quick_sort()
        elif (select == SearchModes.BINARY_SEARCH_THREE.value):
            print("Nothing Implemented\n")
        elif (select == SearchModes.BINARY_SEARCH_FOUR.value):
            print("Nothing Implemented\n")
        elif (select == SearchModes.BINARY_SEARCH_FIVE.value):
             print("Nothing Implemented\n")
       #  Add your own choices here
        else:
            sys.stderr.write("The value %d is not supported\n" % select)
            sys.exit(23)
            
    # You may find this helpful
    # It swaps the element at index a and the element at index b in array
    def swap(self, a, b):
        temp = self.array[a]
        self.array[a] = self.array[b]
        self.array[b] = temp
        
    def insertion_sort(self):
        n = len(self.array)
        for i in range(n):
            key = self.array[i]
            j = i - 1
            while (j >= 0 and self.array[j] > key):
                self.array[j + 1] = self.array[j]
                j = j - 1
            self.array[j + 1] = key
            
        
    # Hint: you probably want to define a help function for the recursive call
    def quick_sort_recursive(self, start, end):
        if (start >= end):
            return
        mid = self.array[end]
        left = start
        right = end - 1
        while (left < right):
            while (self.array[left] < mid and left < right):
                left = left + 1
            while (self.array[right] >= mid and left < right):
                right = right - 1
            self.swap(left, right)
            
        if (self.array[left] >= self.array[end]):
            self.swap(left, end)
        else:
            left = left + 1
        
        if (left > 0):
            self.quick_sort_recursive(start, left - 1)
        self.quick_sort_recursive(left + 1, end)
    
    def quick_sort(self):
        self.quick_sort_recursive(0, len(self.array) - 1)

            
    
            
    
class SearchModes(Enum):
    LINEAR_SEARCH = 0
    BINARY_SEARCH_ONE = 1
    BINARY_SEARCH_TWO = 2
    BINARY_SEARCH_THREE = 3
    BINARY_SEARCH_FOUR = 4
    BINARY_SEARCH_FIVE = 5
    

File: lab3/java/comp26120/darray.java
package comp26120;

import java.util.ArrayList;

public class darray implements set<String> {
    private ArrayList<String> list = new ArrayList<String>();
    private boolean sorted = false;

    SearchModes mode;
    int verbose;

    public darray(speller_config config) {
	mode = SearchModes.getSearchMode(config.mode);
	verbose = config.verbose;
    }
    
    public void insert(String value) {
        list.add(value);
        
        // changing the array means it may no longer be sorted
        sorted = false;
    }
    
    public boolean find(String value) {
        if (mode == SearchModes.LINEAR_SEARCH) {
            // TODO implemented linear search through list
        } else { //Binary Search
            if (! sorted) {
                if (verbose > 0) {
                    System.out.println("Dynamic Array not sorted, sorting... \n");
                }
                
                sort(mode);
                if (verbose > 0) {
                    System.out.println("Dynamic Array sorted\n");
                }
                
                sorted = true;
            }
            // TODO implement binary search through list
        }
	return false;
    }

    public void print_set ()
    {
	System.out.print("DArray:\n");
	for(int i=0;i<list.size();i++){
	    System.out.format("\t%s\n",list.get(i));
	}
    }

    public void print_stats()
    {
	System.out.format("Dynamic array contains %d elements\n",list.size());
    }


    // Sorting Methods

    public enum SearchModes   { LINEAR_SEARCH,  // =0 in mode flag 
				BINARY_SEARCH_ONE,  // = 1
				BINARY_SEARCH_TWO, // = 2
				BINARY_SEARCH_THREE, // =3
				BINARY_SEARCH_FOUR, // = 4
				BINARY_SEARCH_FIVE; // =5


	public static SearchModes getSearchMode(int i) {
	    switch (i) {
	    case 1:
		return BINARY_SEARCH_ONE;
	    case 2:
		return BINARY_SEARCH_TWO;
	    case 3:
		return BINARY_SEARCH_THREE;
	    case 4:
		return BINARY_SEARCH_FOUR;
	    case 5:
		return BINARY_SEARCH_FIVE;
	    default:
		return LINEAR_SEARCH;
		
	    }
	}

				
    };

    private void sort(SearchModes select){

	switch(select){
	case BINARY_SEARCH_ONE   : insertion_sort(); break;
	case BINARY_SEARCH_TWO   : quick_sort(); break;
	case BINARY_SEARCH_THREE :
	case BINARY_SEARCH_FOUR  :
	case BINARY_SEARCH_FIVE  :  // Add your own choices here
	default: 
	    System.err.format("The value %d is not supported in sorting.c\n",
              select);
	    // Keep this exit code as the tests treat 23 specially
	    System.exit(23);
	}
    }


    // You may find this helpful
    // It swaps the element at index a and the element at index b in list
    private void swap(int a, int b)
    {
        String temp = list.get(a);
        list.set(a, list.get(b));
        list.set(b, temp);
    }


    private void insertion_sort(){
	System.err.format("Not implemented\n");
	System.exit(-1);
    }

    // Hint: you probably want to define a help function for the recursive call
    private void quick_sort() {
	System.err.format("Not implemented\n");
	System.exit(-1);
    }

}
